rust/operator/src/lib.rs

```diff
@@ -207,58 +210,66 @@ impl SparkState {
+                        .as_slice(),
+                        &RoleGroupEligibleNodes::from(&self.eligible_nodes),
+                        &PodToNodeMapping::from(&self.existing_pods, None),
+                    )?;
 
-                    for node in nodes_that_need_pods {
-                        let node_name = if let Some(node_name) = &node.metadata.name {
-                            node_name
-                        } else {
-                            warn!("No name found in metadata, this should not happen! Skipping node: [{:?}]", node);
-                            continue;
-                        };
-                        debug!(
-                            "Creating pod on node [{}] for [{}] role and group [{}]",
-                            node.metadata
-                                .name
-                                .as_deref()
-                                .unwrap_or("<no node name found>"),
-                            role_str,
-                            role_group
-                        );
-
-                        // now we have a node that needs pods -> get validated config
+                    let mapping = state.remaining_mapping().get_filtered(role_str, role_group);
+
+                    if let Some((pod_id, node_id)) = mapping.iter().next() {
+                        // now we have a node that needs a pod -> get validated config
                         let validated_config = config_for_role_and_group(
-                            role_str,
-                            role_group,
+                            pod_id.role(),
+                            pod_id.group(),
                             &self.validated_role_config,
                         )?;
 
                         let config_maps = self
-                            .create_config_maps(role_str, role_group, validated_config)
+                            .create_config_maps(pod_id, validated_config, &state.mapping())
                             .await?;
 
                         self.create_pod(
+                            pod_id,
                             &role,
-                            role_group,
-                            node_name,
+                            &node_id.name,
                             &config_maps,
                             validated_config,
                         )
                         .await?;
 
-                        changes_applied = true;
-                    }
+                        history.save(&self.context.resource).await?;
 
-                    // we do this here to make sure pods of each role are started after each other
-                    // roles start in rolling fashion
-                    // pods of each role group start in parallel
-                    // master > worker > history server
-                    if changes_applied {
                         return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
                     }
```

Comment:
Is that still a rolling fashion (e.g. all masters, then all workers...)?

Reply:
The order in which roles are deployed is determined by `SparkRole::iter()` and that does master, worker and history.